### PR TITLE
Add a `zero_tangent_internal` that uses `stackdict` to avoid circular reference induced infinite recursion 

### PR DESCRIPTION
**Pull request by sunxd3**
Address #204

The error we are to address is, e.g.
```julia
mutable struct Foo
    x
    y::Float64
end

foo = Foo(nothing, 1.0)
foo.x = foo # what is the `zero_tangent` of this object?
```

I think it is reasonable to set the tangent type as `{NoTangent(), Float64}`. This is quite subtle as `foo.x === foo` but  `zero_tangent(foo).fields.x !== zero_tangent(foo)`. But I think a recursive base case is required, I am not sure how this may break things.

Following the discussion and suggestions, I added a `stackdict` to track the `zero_tangent`s that has been encountered. This should give the benefits of sharing the same `zero_tangent` when two fields are egal.

### CODE REVIEW

**Comment by willtebbutt** [comment]
A few misc points to aid offline discussion:

#### Arrays

`Array`s can contain circular references. e.g.
```julia
julia> x = Vector{Any}(undef, 1)
1-element Vector{Any}:
 #undef

julia> x[1] = x
1-element Vector{Any}:
 1-element Vector{Any}:#= circular reference @-1 =#
```
so their method of `zero_tangent_internal` will require the `IdDict`.

I think the same is probably true of `SimpleVector`s.

#### Tuples / NamedTuples

`Tuple`s (also `NamedTuple`s) aren't always free of alias / circular-references. Consider, for example,
```julia
julia> x = randn(2)
2-element Vector{Float64}:
 -1.4593203693602337
  1.8727501637988224

julia> y = (x, x)
([-1.4593203693602337, 1.8727501637988224], [-1.4593203693602337, 1.8727501637988224])
```

**Comment by willtebbutt** [comment]
> WIth the Foo example above, the result gradient of foo should have type looks like MutableTangent{NT{typeof(foo), Float64}}

This isn't _quite_ correct. The typeof the gradient of `foo` should be `tangent_type(Foo)`, which is `Tangent{@NamedTuple{x, y::Float64}}`. However, the _value_ of the `x` field of this tangent will be something concrete (e.g. `tangent_type(Foo)`)

File: src/tangents.jl
Comment (sunxd3) [inline src/tangents.jl]: this will make `zero_tangent_internal` type-unstable

Comment (willtebbutt) [reply src/tangents.jl]: Could you try adding a type assertion here? i.e.
```suggestion
        return stackdict[x]::fdata_type(tangent_type(P))
```
ought to work nicely.

File: src/tangents.jl
Comment (willtebbutt) [inline src/tangents.jl]: This could be simplified to
```suggestion
    return isbitstype(P) ? zero_tangent_internal(x) : zero_tangent_internal(x, IdDict())
```

File: src/tangents.jl
Comment (willtebbutt) [inline src/tangents.jl]: Should this maybe live inside the `tangent_type(P) <: MutableTangent` if statement? It'll be correct either way, but where it currently is, I imagine it'll damage the performance if `tangent_type(P) == NoTangent`, or is a `struct`.

File: src/tangents.jl
Comment (willtebbutt) [inline src/tangents.jl]: Should regular immutable `struct`s be put into the `IdDict`? I don't _think_ that they can ever alias one another / contain circular references, but would be happy to be shown wrong on this point!

File: src/tangents.jl
Comment (sunxd3) [reply src/tangents.jl]: The clause with condition `!( tangent_type(P) <: MutableTangent)` is for `struct` that are not bitstype. One example I encountered here is `SubArray`.

for instance,
```julia
struct A
    x
    y
end

x = [1.0, 2.0, 3.0]
it = Tapir.zero_tangent(A(view(x, 1:2), view(x, 1:2))

# returns
Tangent{@NamedTuple{x, y}}(@NamedTuple{x, y}((Tangent{@NamedTuple{parent::Vector{Float64}, indices::NoTangent, offset1::NoTangent, stride1::NoTangent}}((parent = [0.0, 0.0, 0.0], indices = NoTangent(), offset1 = NoTangent(), stride1 = NoTangent())), Tangent{@NamedTuple{parent::Vector{Float64}, indices::NoTangent, offset1::NoTangent, stride1::NoTangent}}((parent = [0.0, 0.0, 0.0], indices = NoTangent(), offset1 = NoTangent(), stride1 = NoTangent())))))

# and
it.fields.x === it.fields.y
```

without saving to stackdict,

returns
```julia
(([0.0, 0.0, 0.0], NoTangent(), NoTangent(), NoTangent()), ([0.0, 0.0, 0.0], NoTangent(), NoTangent(), NoTangent()))
```

This is not necessary for resolving the error, but will have the same tangent for the same object, as you indicated in the github issue description

File: src/tangents.jl
Comment (sunxd3) [reply src/tangents.jl]: similar to the above response

**Comment by willtebbutt** [comment]
It looks like the CI for the inter-AD benchmarks is failing because it doesn't have permission to comment on this PR, as your PR is coming from a fork of this repo. I'm going to make a PR from my own fork which turns this benchmark suite off in this situation.

**Comment by willtebbutt** [comment]
inter-AD benchmarks are now (correctly) being skipped.

File: src/tangents.jl
Comment (willtebbutt) [inline src/tangents.jl]: This method doesn't appear to be used. Could we remove it?

File: src/tangents.jl
Comment (willtebbutt) [inline src/tangents.jl]: Would you consider using diagonal dispatch here, rather than error checking? Something like
```julia
    MutableTangent{Tfields}(fields::Tfields) where {Tfields} = new{Tfields}(fields)
```

File: src/tangents.jl
Comment (willtebbutt) [inline src/tangents.jl]: Could `_map_if_assigned!` be used here, rather than writing out a for loop?

File: src/tangents.jl
Comment (willtebbutt) [inline src/tangents.jl]: I think we can probably get away without doing this here. Since `Tuple`s and `NamedTuple`s are immutable, they don't have an address, so the same reasoning applies to them as does to `struct`s.

So I think you can avoid the check, and you don't ever need to put them in the `stackdict`, so the line just before the end of the function can be dropped as well.

**Review by willtebbutt — COMMENTED** [review]
This is great. I've left direct comments on several details.

Some high-level thoughts:
- do we really need all four inner constructors for `MutableTangent`? It definitely looks like one of them isn't actually used, but I wonder whether others could go? Plainly, we need the partial-initialisation one, but I'm not clear on why we need the others any more.
- would it make sense to update the docstring for `zero_tangent` to tell users who wish to implement `zero_tangent` for their type that they should look at `zero_tangent_internal`?
- why are we calling the `IdDict` variable `stackdict`? In particular, where does the `stack` part of the name come from?

File: src/tangents.jl
Comment (sunxd3) [reply src/tangents.jl]: This is motivated by a test error, e.g.
```julia
julia> MutableTangent{@NamedTuple{a, b::Float64}}((;a = 1.0, b = 2.0))
ERROR: MethodError: no method matching MutableTangent{@NamedTuple{a, b::Float64}}(::@NamedTuple{a::Float64, b::Float64})

Closest candidates are:
  MutableTangent{Tfields}(::Tfields) where Tfields
   @ Main REPL[1]:3
  MutableTangent{Tfields}() where Tfields
   @ Main REPL[1]:4

Stacktrace:
 [1] top-level scope
   @ REPL[6]:1
```
where `@NamedTuple{a, b::Float64}` can be a `tengent_type` return. 


Technically, this should include
```julia
MutableTangent{Tfields}(fields::Tfields) where {Tfields} = new{Tfields}(fields)
```
but I left it there to catch special case.

File: src/tangents.jl
Comment (sunxd3) [reply src/tangents.jl]: agreed, I put it there before to catch aliasing, but as we established, a better way to handle aliasing of struct is taking care of the field that actually require tangents

File: src/tangents.jl
Comment (willtebbutt) [reply src/tangents.jl]: Ahh right, yes, I see why this is needed -- I'm very glad we have a decent number of test cases now!

Perhaps this would work:
```suggestion
    function MutableTangent{Tfields}(fields::NamedTuple{names}) where {names, Tfields<:NamedTuple{names}}
        return new{Tfields}(fields)
    end
```

File: src/tangents.jl
Comment (willtebbutt) [inline src/tangents.jl]: Another thing: do we actually need these two separate paths here? If I'm not mistaken, the second case should cover everything that we need, and be just as performant. I might be wrong though...

**Comment by willtebbutt** [comment]
Huh, I see. I'm happy to follow `deepcopy`'s conventions. Actually, on that note, could you leave a comment somewhere in the source saying that this design follows that of `deepcopy`? It's just helpful context for anyone else who might see this who is already familiar with `deepcopy`.

File: src/tangents.jl
Comment (sunxd3) [reply src/tangents.jl]: yeah, I think so

File: src/tangents.jl
Comment (sunxd3) [reply src/tangents.jl]: that works, thanks for the suggestion

File: src/tangents.jl
Comment (willtebbutt) [inline src/tangents.jl]:
```suggestion
2. Otherwise, `zero_tangent_internal` takes another argument which is an `IdDict`, which
handles both circular references and aliasing correctly.
```

**Review by willtebbutt — APPROVED** [review]
I'm basically happy with this. Just one comment.

Could you please bump the patch version so that we can make a release?

**Comment by willtebbutt** [comment]
Thanks for sorting this. I'll merge when CI passes.

**Comment by willtebbutt** [comment]
Actually, you have merge right's I believe, so feel free to merge when CI passes.

**Comment by willtebbutt** [comment]
Apologies, but I've made the patch version in this PR out of date -- you'll need to bump it again!

**Pull request merged**
